Content.Shared/RCD/Systems/RCDSystem.cs

```diff
@@ -35,312 +41,588 @@ public sealed class RCDSystem : EntitySystem
+                    _adminLogger.Add(LogType.RCD, LogImpact.High, $"{ToPrettyString(user):user} used RCD to delete {ToPrettyString(target):target}");
+                    QueueDel(target);
                 }
-                return true;
-            default:
-                return false; //I don't know why this would happen, but sure I guess. Get out of here invalid state!
+
+                break;
         }
     }
 
-    private void NextMode(EntityUid uid, RCDComponent comp, EntityUid user)
+    #endregion
+
+    #region Utility functions
+
+    public bool TryGetMapGridData(EntityCoordinates location, [NotNullWhen(true)] out MapGridData? mapGridData)
     {
-        _audio.PlayPredicted(comp.SwapModeSound, uid, user);
+        mapGridData = null;
+        var gridUid = location.GetGridUid(EntityManager);
+
+        if (!TryComp<MapGridComponent>(gridUid, out var mapGrid))
+        {
+            location = location.AlignWithClosestGridTile(1.75f, EntityManager);
+            gridUid = location.GetGridUid(EntityManager);
+
+            // Check if we got a grid ID the second time round
+            if (!TryComp(gridUid, out mapGrid))
+                return false;
+        }
+
+        gridUid = mapGrid.Owner;
 
-        var mode = (int) comp.Mode;
-        mode = ++mode % RcdModeCount;
-        comp.Mode = (RcdMode) mode;
-        Dirty(uid, comp);
+        var tile = _mapSystem.GetTileRef(gridUid.Value, mapGrid, location);
+        var position = _mapSystem.TileIndicesFor(gridUid.Value, mapGrid, location);
+        mapGridData = new MapGridData(gridUid.Value, mapGrid, location, tile, position);
 
-        var msg = Loc.GetString("rcd-component-change-mode", ("mode", comp.Mode.ToString()));
-        _popup.PopupClient(msg, uid, user);
+        return true;
     }
 
-    private bool IsTileBlocked(TileRef tile)
+    public bool DoesCustomBoundsIntersectWithFixture(Box2 boundingBox, Angle boundingBoxAngle, EntityUid fixtureOwner, Fixture fixture)
     {
-        return _turf.IsTileBlocked(tile, CollisionGroup.MobMask);
+        var entXformComp = Transform(fixtureOwner);
+        var entXform = new Transform(entXformComp.LocalPosition, entXformComp.LocalRotation);
+
+        var verts = new ValueList<Vector2>()
+        {
+            boundingBox.BottomLeft,
+            boundingBox.BottomRight,
+            boundingBox.TopRight,
+            boundingBox.TopLeft
+        };
+
+        var poly = new PolygonShape();
+        poly.Set(verts.Span, 4);
+
+        var polyXform = new Transform(entXformComp.LocalPosition, (float) boundingBoxAngle);
+        return poly.ComputeAABB(polyXform, 0).Intersects(fixture.Shape.ComputeAABB(entXform, 0));
+    }
+
+    public void UpdateCachedPrototype(EntityUid uid, RCDComponent component)
+    {
+        if (component.ProtoId.Id != component.CachedPrototype?.Prototype)
+            component.CachedPrototype = _protoManager.Index(component.ProtoId);
+    }
+
+    #endregion
+}
+
```

Comment:
This really shouldn't be public, callers shouldn't have to worry about this kind of stuff.

Also for reference the 2 Transforms you do at the top here are more expensive than just indexing the cached prototype so I really don't think it's worth it. If you want to save performance add in EntityQuery<T> around the place.

Reply:
I've moved a lot of the cachable data in DoesCustomBoundsIntersectWithFixture to the RCD component  and the cached prototype. I'm not too worried about performance, as it only triggers with thin entities, like windows

UpdateCachedPrototype will need to be public for prediction purposes